File: MCPServer/TodosMcpTool.cs
using MCPServer.Data;
using System.ComponentModel;
using ModelContextProtocol.Server;
using Microsoft.EntityFrameworkCore;

namespace MCPServer
{
    [McpServerToolType]
    public class TodosMcpTool
    {
        private readonly DatabaseContext _db;

        public TodosMcpTool(DatabaseContext db)
        {
            _db = db;
        }

        [McpServerTool, Description("Creates a new todo with a description and creation date.")]
        public async Task<string> CreateTodoAsync(
            [Description("Description of the todo")] string description,
            [Description("Creation date of the todo")] DateTime createdDate)
        {
            var todo = new Todo
            {
                Description = description,
                CreatedDate = createdDate
            };
            _db.Todos.Add(todo);
            await _db.SaveChangesAsync();
            return $"Todo created: {todo.Description} (Id: {todo.Id})";
        }

        [McpServerTool, Description("Reads all todos, or a single todo if an id is provided.")]
        public async Task<List<Todo>> ReadTodosAsync(
            [Description("Id of the todo to read (optional)")] string? id = null)
        {
            if (!string.IsNullOrWhiteSpace(id) && int.TryParse(id, out int todoId))
            {
                var todo = await _db.Todos.FindAsync(todoId);
                if (todo == null) return new List<Todo>();
                return new List<Todo> { todo };
            }
            var todos = await _db.Todos.OrderBy(t => t.Id).ToListAsync();
            return todos;
        }

        [McpServerTool, Description("Updates the specified todo fields by id.")]
        public async Task<string> UpdateTodoAsync(
            [Description("Id of the todo to update")] string id,
            [Description("New description (optional)")] string? description = null,
            [Description("New creation date (optional)")] DateTime? createdDate = null)
        {
            if (!int.TryParse(id, out int todoId))
                return "Invalid todo id.";
            var todo = await _db.Todos.FindAsync(todoId);
            if (todo == null) return $"Todo with Id {todoId} not found.";
            if (!string.IsNullOrWhiteSpace(description)) todo.Description = description;
            if (createdDate.HasValue) todo.CreatedDate = createdDate.Value;
            await _db.SaveChangesAsync();
            return $"Todo {todo.Id} updated.";
        }

        [McpServerTool, Description("Deletes a todo by id.")]
        public async Task<string> DeleteTodoAsync(
            [Description("Id of the todo to delete")] string id)
        {
            if (!int.TryParse(id, out int todoId))
                return "Invalid todo id.";
            var todo = await _db.Todos.FindAsync(todoId);
            if (todo == null) return $"Todo with Id {todoId} not found.";
            _db.Todos.Remove(todo);
            await _db.SaveChangesAsync();
            return $"Todo {todo.Id} deleted.";
        }
    }
}

File: go/internal/tools/todos_mcp_tool.go
package tools

import (
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/matpadley/MCPServer_Demo/go/internal/data"
)

// TodosMcpTool provides MCP tools for todo management
type TodosMcpTool struct {
	db *data.DatabaseContext
}

// NewTodosMcpTool creates a new TodosMcpTool instance
func NewTodosMcpTool(db *data.DatabaseContext) *TodosMcpTool {
	return &TodosMcpTool{db: db}
}

// CreateTodoAsync creates a new todo with a description and creation date
func (t *TodosMcpTool) CreateTodoAsync(description string, createdDate time.Time) (string, error) {
	todo, err := t.db.CreateTodoAsync(data.CreateTodoInput{
		Description: description,
		CreatedDate: createdDate,
	})
	if err != nil {
		return "", fmt.Errorf("error creating todo: %w", err)
	}

	return fmt.Sprintf("Todo created: %s (Id: %d)", *todo.Description, todo.ID), nil
}

// ReadTodosAsync reads all todos, or a single todo if an id is provided
func (t *TodosMcpTool) ReadTodosAsync(id *string) ([]data.Todo, error) {
	if id != nil && strings.TrimSpace(*id) != "" {
		todoID, err := strconv.Atoi(strings.TrimSpace(*id))
		if err != nil {
			// Invalid ID, return empty list like other implementations
			return []data.Todo{}, nil
		}
		return t.db.ReadTodosAsync(todoID)
	}

	return t.db.ReadTodosAsync()
}

// UpdateTodoAsync updates the specified todo fields by id
func (t *TodosMcpTool) UpdateTodoAsync(id string, description *string, createdDate *time.Time) (string, error) {
	todoID, err := strconv.Atoi(id)
	if err != nil {
		return "Invalid todo id.", nil
	}

	updateInput := data.UpdateTodoInput{}
	if description != nil && strings.TrimSpace(*description) != "" {
		updateInput.Description = description
	}
	if createdDate != nil {
		updateInput.CreatedDate = createdDate
	}

	updated, err := t.db.UpdateTodoAsync(todoID, updateInput)
	if err != nil {
		return "", fmt.Errorf("error updating todo: %w", err)
	}

	if !updated {
		return fmt.Sprintf("Todo with Id %d not found.", todoID), nil
	}

	return fmt.Sprintf("Todo %d updated.", todoID), nil
}

// DeleteTodoAsync deletes a todo by id
func (t *TodosMcpTool) DeleteTodoAsync(id string) (string, error) {
	todoID, err := strconv.Atoi(id)
	if err != nil {
		return "Invalid todo id.", nil
	}

	deleted, err := t.db.DeleteTodoAsync(todoID)
	if err != nil {
		return "", fmt.Errorf("error deleting todo: %w", err)
	}

	if !deleted {
		return fmt.Sprintf("Todo with Id %d not found.", todoID), nil
	}

	return fmt.Sprintf("Todo %d deleted.", todoID), nil
}